clients/src/test/java/org/apache/kafka/common/network/EchoServer.java

```diff
@@ -71,49 +72,57 @@ public void renegotiate() {
-                                if (renegotiate.get()) {
-                                    renegotiate.set(false);
-                                    ((SSLSocket) socket).startHandshake();
-                                }
-                                byte[] bytes = new byte[size];
-                                input.readFully(bytes);
-                                output.writeInt(size);
-                                output.write(bytes);
-                                output.flush();
-                            }
-                        } catch (IOException e) {
-                            // ignore
-                        } finally {
+                synchronized (sockets) {
+                    if (closing) {
+                        break;
+                    }
+                    sockets.add(socket);
+                    Thread thread = new Thread() {
+                        @Override
+                        public void run() {
                             try {
-                                socket.close();
+                                DataInputStream input = new DataInputStream(socket.getInputStream());
+                                DataOutputStream output = new DataOutputStream(socket.getOutputStream());
+                                while (socket.isConnected() && !socket.isClosed()) {
+                                    int size = input.readInt();
+                                    if (renegotiate.get()) {
+                                        renegotiate.set(false);
+                                        ((SSLSocket) socket).startHandshake();
+                                    }
+                                    byte[] bytes = new byte[size];
+                                    input.readFully(bytes);
+                                    output.writeInt(size);
+                                    output.write(bytes);
+                                    output.flush();
+                                }
                             } catch (IOException e) {
                                 // ignore
+                            } finally {
+                                try {
+                                    socket.close();
+                                } catch (IOException e) {
+                                    // ignore
+                                }
                             }
                         }
-                    }
-                };
-                thread.start();
-                threads.add(thread);
+                    };
+                    thread.start();
+                    threads.add(thread);
+                }
             }
         } catch (IOException e) {
             // ignore
         }
     }
 
     public void closeConnections() throws IOException {
-        for (Socket socket : sockets)
-            socket.close();
+        synchronized (sockets) {
+            closing = true;
```

Comment:
`closing` should be set in `close()` rather than `closeConnections` - i.e. set in when closing server socket, not when closing client connections.